### PR TITLE
declarative Online DDL

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

This PR introduces (partial) declarative schema changes in Online DDL. In a declarative schema change you tell vitess "this is a table's schema, make it so". If the table does not exist, vitess creates it. If the table exists and with a different schema, vitess evaluates the required `ALTER` statement and migrates the table.

Declarative changes have the property of being _idempotent_. You may issue the same migration multiple times with no harm done. If vitess find that no change is required, it implicitly markes the migration as successful.

This PR implements _partial_ declarative changes. In a _full_ declarative schema change, one gives their database a list of all tables in a schema. In a partial declarative schema change, we give vitess a _single_ table at a time.

This means:

- If you want to create a table, ask vitess to `CREATE TABLE`
- If you want to modify a table, ask vitess to `CREATE TABLE`
- If you want to drop a table, ask vitess to `DROP TABLE`.

Vitess supports the `-declarative` flag in `dddl_strategy` to enable this functionality.

### Usage

consider:
```sql
set @@ddl_strategy='online -declarative';
create table t(id int primary key); -- table created
create table t(id int primary key, t text); -- table modified
create table t(id int primary key, t text); -- no changes, implicitly successful
create table t(id int primary key, t text, key t_idx(t(32))); -- table modified
drop table t; -- table dropped
drop table t; -- no changes, implicitly successful
```

you may use `-declarative` in all online DDL strategies: `online`, `gh-ost`, `pt-osc`.

Documentation to follow.

## Related Issue(s)

Tracking: #6926 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Reviewers: the main changes to review are in `executor.go`:

- new `evaluateDeclarativeDiff()` function
- within `executeMigration()` function

when we handle a new migration request, we now first check to see if it's `declarative`. If so, we may end up actually running a modified migration than defined. We break down into the following cases:

- it's a `revert` (no special behavior)
- `alter` is not allowed
- `drop`: silently mark as successful if the table does not exist. Otherwise treat that as a normal `drop`
- `create`: either
  - the table does not exist -> proceed as normal `create`
  - the table exists -> use `tengo` to evaluate the diff statement from existing table into desired table. Either:
    - there is no diff (exact same schema): silently mark as successful
    - there is a diff: modify into an `alter` statement, to be handled later on.

Diffing the table via `tengo`:
- `tengo` is a library which uses `sqlx` to connect to to the database
- we have slightly modified `tengo`, https://github.com/planetscale/tengo
- we create an account and connection for `sqlx` & `tengo`
- we create a temporary table within the tablet server's schema (keyspace), which has the new design
- we diff a the specific original table with this temporary table
  - `tengo` doesn't like diffing tables of different names, because it assumes identical names in different schemas. We hack around by rewriting `tengo`'s analysis of the table
- the diff is a SQL string, which is the `ALTER` options (e.g `add column i int, drop key k`), or empty if the tables are identical

**Comment by mavenraven** [comment]
@shlomi-noach Awesome! I'm not going to review the implementation unless you'd like me to. From a perspective of a user of the functionality, it looks perfect.

File: go/vt/vttablet/onlineddl/executor.go [hunk 1]
```diff
@@ -1473,6 +1496,129 @@ func (e *Executor) executeRevert(ctx context.Context, onlineDDL *schema.OnlineDD
...
+
+	return alterClause, nil
+}
+
+// executeMigration executes a single migration. It analyzes the migration type:
+// - is it declarative?
+// - is it CREATE / DROP / ALTER?
+// - it is a Revert request?
+// - what's the migration strategy?
+// The function invokes th eappropriate handlers for each of those cases.
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/onlineddl/executor.go:1621]: th eappropriate handlers => the appropriate handlers

File: go/vt/vttablet/onlineddl/executor.go [hunk 2]
```diff
@@ -1487,6 +1633,78 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
...
+			// - Implicitly do nothing, if the table does not exist
+
+			exists, err := e.tableExists(ctx, onlineDDL.Table)
+			if err != nil {
+				return failMigration(err)
+			}
+			if exists {
+				// table does exist, so this declarative DROP turns out to really be an actual DROP. No further action is needed here
+			} else {
+				// table does not exist. We mark this CREATE as implicitly sucessful
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/onlineddl/executor.go:1655]: should be DROP here and not CREATE?

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm. This is really nice!

File: go/vt/vttablet/onlineddl/executor.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go]: fixed!

File: go/vt/vttablet/onlineddl/executor.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go]: fixed

**Pull request merged**